File: src/models/Reputation.js
const mongoose = require('mongoose');

const reputationSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'L\'utilisateur est requis'],
    unique: true,
    index: true
  },
  score: {
    type: Number,
    default: 0,
    min: [0, 'Le score ne peut pas être négatif'],
    max: [100, 'Le score ne peut pas dépasser 100']
  },
  rideCount: {
    type: Number,
    default: 0,
    min: [0, 'Le nombre de balades ne peut pas être négatif']
  },
  punctualityScore: {
    type: Number,
    default: 50,
    min: [0, 'Le score de ponctualité ne peut pas être négatif'],
    max: [100, 'Le score de ponctualité ne peut pas dépasser 100']
  },
  cancellationRate: {
    type: Number,
    default: 0,
    min: [0, 'Le taux d\'annulation ne peut pas être négatif'],
    max: [100, 'Le taux d\'annulation ne peut pas dépasser 100']
  },
  feedbackCount: {
    type: Number,
    default: 0,
    min: [0, 'Le nombre de feedbacks ne peut pas être négatif']
  },
  level: {
    type: String,
    enum: ['bronze', 'silver', 'gold', 'platinum'],
    default: 'bronze',
    index: true
  }
}, {
  timestamps: true
});

// Index pour améliorer les performances
reputationSchema.index({ score: -1 });
reputationSchema.index({ level: 1, score: -1 });

// Méthode pour mettre à jour le niveau basé sur le score
reputationSchema.methods.updateLevel = function() {
  if (this.score >= 90) {
    this.level = 'platinum';
  } else if (this.score >= 75) {
    this.level = 'gold';
  } else if (this.score >= 60) {
    this.level = 'silver';
  } else {
    this.level = 'bronze';
  }
};

// Middleware pre-save pour mettre à jour le niveau
reputationSchema.pre('save', function(next) {
  if (this.isModified('score')) {
    this.updateLevel();
  }
  next();
});

module.exports = mongoose.model('Reputation', reputationSchema);

File: flutter_app/lib/models/reputation.dart
class Reputation {
  final String id;
  final String userId;
  final int score; // Score de réputation (0-100)
  final int rideCount;
  final int punctualityScore; // Score de ponctualité (0-100)
  final double cancellationRate; // Taux d'annulation (0-1)
  final int feedbackCount;
  final String level; // bronze, silver, gold, platinum

  Reputation({
    required this.id,
    required this.userId,
    required this.score,
    required this.rideCount,
    required this.punctualityScore,
    required this.cancellationRate,
    required this.feedbackCount,
    required this.level,
  });

  factory Reputation.fromJson(Map<String, dynamic> json) {
    return Reputation(
      id: json['id'] ?? json['_id'] ?? '',
      userId: json['userId'] is Map
          ? (json['userId']['_id'] ?? json['userId']['id'] ?? '').toString()
          : json['userId']?.toString() ?? '',
      score: json['score'] ?? 0,
      rideCount: json['rideCount'] ?? 0,
      punctualityScore: json['punctualityScore'] ?? 50,
      cancellationRate: (json['cancellationRate'] ?? 0).toDouble(),
      feedbackCount: json['feedbackCount'] ?? 0,
      level: json['level'] ?? 'bronze',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'score': score,
      'rideCount': rideCount,
      'punctualityScore': punctualityScore,
      'cancellationRate': cancellationRate,
      'feedbackCount': feedbackCount,
      'level': level,
    };
  }

  String get levelDisplayName {
    switch (level) {
      case 'bronze':
        return 'Bronze';
      case 'silver':
        return 'Argent';
      case 'gold':
        return 'Or';
      case 'platinum':
        return 'Platine';
      default:
        return 'Bronze';
    }
  }
}












